dns/zone.py

```diff
@@ -82,6 +82,38 @@ class DigestVerificationFailure(dns.exception.DNSException):
     """The ZONEMD digest failed to verify."""
 
 
+def _validate_name(
+    name: dns.name.Name,
+    origin: Optional[dns.name.Name],
+    relativize: bool,
+) -> dns.name.Name:
+    # This name validation code is shared by Zone and Version
+    if origin is None:
+        # This should probably never happen as other code (e.g.
+        # _rr_line) will notice the lack of an origin before us, but
+        # we check just in case!
+        raise KeyError("no zone origin is defined")
+    if name.is_absolute():
+        if not name.is_subdomain(origin):
+            raise KeyError("name parameter must be a subdomain of the zone origin")
+        if relativize:
+            name = name.relativize(origin)
+    elif not relativize:
+        # We have a relative name in a non-relative zone, so derelativize.
+        name = name.derelativize(origin)
```

Comment:
The new code below converts the `NameTooLong` into `KeyError`, but this code doesn't do that.  Should it?

Reply:
It should.  I pushed a corrected and even simpler bit of code.